lib/stores/S3Store.js

```diff
@@ -505,23 +505,44 @@ class S3Store extends DataStore {
             this._getMetadata(file_id)
                 .then((metadata) => {
                     return this._retrieveParts(file_id)
-                        .then((parts) => {
-                            return {
-                                parts,
-                                metadata,
-                            };
-                        })
+                        .then((parts) => ({ parts, metadata }))
                         .catch((err) => {
-                            throw err;
+                            // getOffset is called from both the HEAD request and the PATCH request
+                            // (via the _processUpload method). The HEAD request calls it (implicitly)
+                            // with with_parts: false, the _processUpload method calls it (explicitly)
+                            // with with_parts: true. The _processUpload method expects upload parts to be returned,
+                            // which would not work when the upload is completed, since there are no parts anymore.
+                            // TODO: refactor to separate these concerns.
+                            // TODO: write a test for this case.
+                            if (with_parts === true || err.code !== 'NoSuchUpload') {
+                                throw err;
+                            }
+                            return this.client.headObject({ Bucket: this.bucket_name, Key: file_id })
+                                .promise()
+                                .then(() => {
+                                    resolve({
+                                        ...this.cache[file_id].file,
+                                        size: metadata.file.upload_length,
+                                        upload_length: metadata.file.upload_length,
+                                    });
+                                    return { isFinished: true };
+                                })
+                                .catch((headError) => {
+                                    throw headError;
+                                });
                         });
                 })
                 .then((data) => {
+                    if (data.isFinished) {
+                        return;
+                    }
```

Comment:
Together with the above change. (Also I think it's fine to `return resolve`, doesn't necessarily need to be separate).
```suggestion
                .then(({ data, isFinished }) => {
                    if (isFinished) {
                        return resolve(data);
                    }
```